monitoring/alerts.rb

```diff
@@ -0,0 +1,216 @@
+  policies = client.list_alert_policies project_name, filter: filter
+
+  policies.each do |policy|
+    if policy.enabled.value == enable
+      p "Policy #{policy.name} is already #{policy.enabled.value ? "enabled" : "disabled"}"
+    else
+      policy.enabled.value = enable
+      update_mask = Google::Protobuf::FieldMask.new paths: ["enabled"]
+      # client.update_alert_policy policy, update_mask
+      p "#{enable ? "Enabled" : "Disabled"} #{policy.name}"
+    end
+  end
+  # [END monitoring_alert_enable_policies]
+end
+
+def list_notification_channels project_id:
+  # [START monitoring_alert_list_channels]
+  client = Google::Cloud::Monitoring::NotificationChannel.new
+  project_name = Google::Cloud::Monitoring::V3::NotificationChannelServiceClient.project_path project_id
+  channels = client.list_notification_channels project_name
+  channels.each do |channel|
+    p channel
+  end
+  # [END monitoring_alert_list_channels]
+end
+
+def replace_notification_channels project_id:, alert_policy_id:, channel_ids: []
+  # [START monitoring_alert_replace_channels]
+  alert_client = Google::Cloud::Monitoring::AlertPolicy.new
+  channel_client = Google::Cloud::Monitoring::NotificationChannel.new
+  project_name = Google::Cloud::Monitoring::V3::AlertPolicyServiceClient.project_path project_id
+
+  policy = Google::Monitoring::V3::AlertPolicy.new(
+    name: Google::Cloud::Monitoring::V3::AlertPolicyServiceClient.alert_policy_path(project_id, alert_policy_id)
+  )
+
+  channel_ids.each do |channel_id|
+    policy.notification_channels <<
+      Google::Cloud::Monitoring::V3::NotificationChannelServiceClient.notification_channel_path(
+        project_id,
+        channel_id
+      )
+  end
+
+  update_mask = Google::Protobuf::FieldMask.new paths: ["notification_channels"]
+  updated_policy = alert_client.update_alert_policy policy, update_mask
+  p "Updated #{updated_policy.name}"
+  # [END monitoring_alert_replace_channels]
+end
+
+def backup project_id:
+  # [START monitoring_alert_backup_policies]
+  alert_client = Google::Cloud::Monitoring::AlertPolicy.new
+  channel_client = Google::Cloud::Monitoring::NotificationChannel.new
+  project_name = Google::Cloud::Monitoring::V3::AlertPolicyServiceClient.project_path project_id
+
+  policies = alert_client.list_alert_policies(project_name).map(&:to_h)
+  channels = channel_client.list_notification_channels(project_name).map(&:to_h)
+  record = {
+    project_id: project_id,
+    policies: policies,
+    channels: channels
+  }
+
+  File.write("backup.json", JSON.pretty_generate(record))
+  puts "Backed up alert policies and notification channels to 'backup.json'."
+  # [END monitoring_alert_backup_policies]
+end
+
+def restore project_id:
+  # [START monitoring_alert_restore_policies]
+  # [START monitoring_alert_create_policy]
+  # [START monitoring_alert_create_channel]
+  # [START monitoring_alert_update_channel]
```

Comment:
I know we wouldn't be the only language to do it this way, but it seems weird having a bunch of region tags start and end  in sync. Is there a reasonable way to break it up, making each region tag cover just the specific snippet it corresponds to?

Reply:
Working on it.

Reply:
@TheRoyalTnetennba added separate examples